worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,134 @@
-    ))
-  } else {
-    logger.info("No job store found in config")
-    None
-  }
-
-  def getConnection = provider.get.getConnection
+  private val jdbcURL = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".URL") else null
+  private val detectIpAddressFailover = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".ipFailover") == "true" else false
+  // Removes "jdbc:mysql://" prefix and ":{port}..." suffix
+  private val dataSourceHostname = if (jdbcURL != null) jdbcURL.replace("jdbc:mysql://", "").split(":")(0) else null
+
+  // Class for storing the ip address of the host, along with an expiration date
+  private case class CachedIpWithExpiration(ip: String, expiration: Long)
+  // Cache for ip address and its expiration for a host
+  private val cachedIpWithExpiration: LinkedHashMap[String, CachedIpWithExpiration] = new LinkedHashMap()
+  // Time (in milliseconds) that the in-memory cache will retain the ip address
+  private val cachedIpTTL: Long = 1000
+  private val getIpNumRetries = 10
+
+  private val pool: Pool = new Pool(getIP)
+
+  // Intended to be used only for tests. This mocks an IP failover every time a connection is retreived
+  private val causeFailoverEveryConnection = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".causeFailoverEveryConnection") == "true" else false
+
+  def createNewConnectionProvider(): Option[HikariCpPoolingConnectionProvider] = {
+    if(dataSource != null) {
+      Some(new HikariCpPoolingConnectionProvider(
+        props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
+        jdbcURL,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
+      ))
+    } else {
+      logger.info("No job store found in config to get connections")
+      None
+    }
+  }
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def hasIpAddressChanged(pool: Pool, dnsIP: String): Boolean = {
+    causeFailoverEveryConnection || pool.ip != dnsIP
+  }
+
+  @tailrec
+  private def retryGettingIp(n: Int)(fn: => String): String = {
+    try {
+      return fn
```

Comment:
Note: In scala 3 this will need to be replaced by a `boundary` instead of a return.

Reply:
There's a lot of things we would need to update for the scala3 upgrade when we get there. Good to point out though